pkg/placement/leadership.go

```diff
@@ -31,85 +32,98 @@ const barrierWriteTimeout = 2 * time.Minute
 // expected to do, so we must react to changes
 //
 // reference: https://github.com/hashicorp/consul/blob/master/agent/consul/leader.go
-func (p *Service) MonitorLeadership(parentCtx context.Context) error {
-	if p.closed.Load() {
-		return errors.New("placement is closed")
-	}
-
-	ctx, cancel := context.WithCancel(parentCtx)
-	p.wg.Add(1)
-	go func() {
-		defer p.wg.Done()
-		defer cancel()
-		select {
-		case <-parentCtx.Done():
-		case <-p.closedCh:
-		}
-	}()
+func (p *Service) MonitorLeadership(ctx context.Context) error {
+	var leaderLoopWg sync.WaitGroup
+	var leaderCtx context.Context
+	var leaderCancelFn context.CancelFunc
 
 	raft, err := p.raftNode.Raft(ctx)
 	if err != nil {
 		return err
 	}
-	var (
-		leaderCh          = raft.LeaderCh()
-		leaderLoopRunning atomic.Bool
-		loopNotRunning    = make(chan struct{})
-	)
-	close(loopNotRunning)
-
-	var leaderLoopCancel context.CancelFunc
-	var leaderCtx context.Context
+	raftLeaderCh := raft.LeaderCh()
+
+	startLeaderLoop := func() {
+		if leaderCtx != nil {
+			log.Error("attempted to start the leader loop while running", p.raftNode.GetID())
+			return
+		}
+
+		leaderCtx, leaderCancelFn = context.WithCancel(ctx)
+
+		leaderLoopWg.Add(1)
+
+		go func(ctx context.Context) {
+			defer leaderLoopWg.Done()
+			err := p.leaderLoop(ctx)
+			if err != nil {
+				log.Error("placement leader loop failed", "error", err)
```

Comment:
Same here

Reply:
Logically, we should be able to ignore this error. We would only get it if `p.raftNode.Raft(ctx)` returns an error, which only happens if the underlying raft server is not ready in time. But we would have never gotten the isLeader signal in the first palce if the raft server wasn't ready.